### DISCUSSION TITLE
Difficulty creating a ADMatReactionNodalKernel class to use materials to define reaction rates

### DISCUSSION THREAD

**Discussion by simopier** [idaholab/TMAP8#9] [Q&A] [answered]
**Initial question:**

'I have been trying to create the ADMatReactionNodalKernel class because I think it would be very useful for TMAP8, but I am running into an issue that I have not been able to fix.
 
I updated conda and moose both this morning again, but when I try to ‘make’ TMAP8 with the new files, I get the following error:
 
Compiling C++ (in opt mode) /Users/simopa/projects/TMAP8/build/unity_src/nodal_kernels_Unity.C...
In file included from /Users/simopa/projects/TMAP8/build/unity_src/nodal_kernels_Unity.C:2:
/Users/simopa/projects/TMAP8/src/nodal_kernels/ADMatReactionNodalKernel.C:29:10: error: use of undeclared identifier 'getADMaterialProperty'; did you mean 'ADMaterialProperty'?
    _mob(getADMaterialProperty<Real>("mob_name"))
         ^
/Users/simopa/projects/moose/framework/build/header_symlinks/MaterialProperty.h:339:7: note: 'ADMaterialProperty' declared here
class ADMaterialProperty : public MaterialPropertyBase<T, true>
      ^
1 error generated.
make: *** [/Users/simopa/projects/TMAP8/build/unity_src/nodal_kernels_Unity.x86_64-apple-darwin13.4.0.opt.lo] Error 1
make: *** Waiting for unfinished jobs....
 
 
I do not understand how to solve it. Looking at the ADMatReaction kernel, it also uses ‘getADMaterialProperty’ and I don’t see why it should be different here. Any suggestions?

Thank you!'

**Comment by simopier** [comment, marked as answer]
**Answer:**

'Nodal kernels won’t work with the current material system unfortunately. Materials are associated with subdomains, e.g. volumes. You can imagine that at an interface between materials, unless you specify an element side, the value of a material property can be discontinuous and worse undefined. What value for the property should be used? It could be even worse for a node: you could be at the corner of 8 subdomains for a hex mesh!

If it’s an expression you have, and your property is not a function of the nonlinear degrees of freedom, why not use a function parameter to keep the expression in the NodalKernel generic? The AD system can still handle the derivatives of that for you.
 
For the general question of kernel vs nodal kernel: let’s use the correct mathematical solution. If we don’t have spatial derivatives I think we should do nodal kernels. And as to the practical challenge of how we implement it (a function, a material, hard-coded in the kernel) let’s just do what’s quickest and easiest for now and I think in the future we will have better solutions (although a function still could be a great solution even going forward depending on your use case). For instance I have plans for doing on-demand property evaluation at any point in the domain. This would be a viable solution even when your property is a function of the nonlinear degrees of freedom.'

**Reply by simopier** [reply]
@lindsayad , I was wondering what you meant when you mentioned:
> nonlinear degrees of freedom

Do you refer to the variables of the system, such as the concentration of tritium?

In that case, that would be an issue because the reaction rates (which would then be described using a function), often depend on the tritium concentration. Until you develop the new capability you mention, I think I would have to hard-code the reaction rate in the kernel.

Let me know if I got it right or if I am missing something.

Thank you!

**Reply by lindsayad** [reply]
@simopier, when @cticenhour made that suggestion, in my head I was imagining that the `Function` would describe a rate constant 'k' which often depends on things like temperature (and parameters like activation energy). I definitely imagine that the reaction *rate* is a function of the rate constant 'k' and the reactant concentrations which would likely have to be pulled in using APIs like `adCoupledValue`. Does the rate constant 'k' depend on the concentrations as well? Shouldn't for an elementary reaction but maybe you have some other scenario. If so, please outline the math behind the reaction rate and then maybe I can be more helpful.

**Reply by simopier** [reply]
@lindsayad , thanks for coming back to me on this! Here is an example of what I would like to do in more details:

Some describe one of the surface reactions as
T<sub>2</sub> + H<sub>2</sub>O &rarr; HTO + HT.

It's effect on the T<sub>2</sub> concentration, [T<sub>2</sub>], can be modeled as
d[T<sub>2</sub>]/dt = -R,
where [T<sub>2</sub>] is the concentration of T<sub>2</sub> and 
R = k [T<sub>2</sub>] (2[T<sub>2</sub>] + [HTO] + [HT]),
where the rate constant k is, as you mentioned, often described as a Arrhenius equation of the temperature. However, in this example, the reaction rate R is also a function of the concentration of the different species concentration, which would all be Variables or AuxVariables of the system.

Since the reaction rates can depend on the species concentrations in different ways, I was hoping to use a nodal kernel that could take in the reaction rate R, which would be defined elsewhere in the input file. That would prevent us from having to write different kernels for different expressions of reaction rates. I originally planned to define R as a material property, but you pointed out the problem with that approach. 

Once solution would still be to write one nodalkernal class for each type of reaction rate, but that would probably be a bit tedious. 

Let me know if you have any suggestions or questions. 
Thanks!

**Reply by lindsayad** [reply]
Ok very clear outline of the problem! How you *want* to approach the problem is definitely the right way ... as you say it would/will be tedious to define a different reaction kernel for every type of reaction, e.g. ternary, binary, unary, quaternary, etc. However, without the availability of `Materials` to us, which is really the only system that allows us to take full advantage of the AD system through `DerivativeParsedMaterial`, I don't know of a better approach than to create `Unary, Binary, Ternary, Quaternary` nodal kernels. Hopefully in your example above you could achieve what you want for say the consumption of the T2 variable using one `Ternary`, and two `Binary` kernels given that those kernels take 3 and 2 coupled variables respectively. Doing reaction networks in this way will be annoying but I think its the simplest path forward right now until we get the on-the-fly-material evaluation ready to go.

Maybe @keniley1 has better ideas though? I believe his application [CRANE](https://github.com/lcpp-org/crane) builds reaction networks through actions but I don't know what his actual residual/Jacobian objects are...I believe he was doing 0D so maybe scalar kernels?

**Reply by keniley1** [reply]
Yeah, for a system of ODEs with no spatial dependence CRANE uses scalar variables and kernels, with rate coefficients stored in AuxScalarVariables. It's not ideal because Jacobians are not handled properly, but I've been able to run pretty large reaction networks using solve_type = linear. (That's inexplicable to me since some of these problems are nonlinear, but it works.)

But to be honest I hardly use 0D; I spent the vast majority of my time and effort on 1D problems. For those the code just uses (AD) materials to store the rate coefficients and adds kernels for each individual reaction. I avoided NodalKernels for the exact same reason you're running into here -- I wanted to use Material properties so I could compute the jacobians easily. In addition, some of the reaction rates require variable gradients because they depend on the flux of one of the species. 

The action in Crane takes each reaction (in this case T2 + H2O → HTO + HT), parses the string into reactants and products, and automatically constructs the source and sink terms for each species.  This does what you want to avoid, though, because each species has a new kernel added for every reaction it's involved in. It looks like you are trying to create a single kernel that calculates the total reaction rate for that species; is that right?

On a side note, right now I only go up to ternary reactions so I annoyingly have kernels for "OneBody", "TwoBody", and "ThreeBody" reactions. It would be easy to make a single kernel for an arbitrary number of reactants using a for loop like this:

```
ADReal val = 1;
for (unsigned int i = 0; i < _num_reactants; ++i)
{
    val *= (*_reactant[i])[_qp]
}

return -_test[_i][_qp] * _rate_coefficient[_qp] * val;

```

though I don't know if that's efficient computationally.

**Reply by lindsayad** [reply]
> The action in Crane takes each reaction (in this case T2 + H2O → HTO + HT), parses the string into reactants and products, and automatically constructs the source and sink terms for each species. This does what you want to avoid, though, because each species has a new kernel added for every reaction it's involved in. It looks like you are trying to create a single kernel that calculates the total reaction rate for that species; is that right?

I think if we made an automated action process, we wouldn't really care how we compose this, e.g. one kernel per reaction or one kernel per species. The general MOOSE mantra is to be as modular as possible so by that criterion one kernel per reaction is the way to go.

@keniley1 ignoring for the moment that you're ultimately constructing `Kernels`, do you think your action parsing of the input file is general enough to be useful to us? Let's imagine that we somehow have a parameter that controls whether you create nodal kernels or kernels depending on the user's desire.

**Reply by lindsayad** [reply]
And let's also pretend for the moment that we have materials that work for both kernels and nodal kernels.

**Reply by keniley1** [reply]
>  @keniley1 ignoring for the moment that you're ultimately constructing Kernels, do you think your action parsing of the input file is general enough to be useful to us? Let's imagine that we somehow have a parameter that controls whether you create nodal kernels or kernels depending on the user's desire.

Yeah, at its most basic level Crane's action is just a string parser. It takes a string like this: 

`e + Ar -> e + e + Ar+     :   {1e-9*Te^(-0.5)}`

And splits that into individual reactions, reactants, products, how many of each are produced/destroyed in each reaction, and rate coefficients. (The rate coefficient is separated from the reaction with the ':' character.)  That's all done very messily in [ChemicalReactionsBase](https://github.com/lcpp-org/crane/blob/master/src/actions/ChemicalReactionsBase.C).  All of that information is very general and doesn't depend on the mesh, variable types, (Nodal/Scalar)Kernels, etc at all.

**Reply by lindsayad** [reply]
Hmm. @cpgr do you do reaction parsing with any custom syntax in the `chemical_reactions` module? I'm thinking about how we can help Moose downstream applications the most here. I'm wondering whether it makes any sense to port the parser capability of @keniley1 into the `chemical_reactions` module. Even unary, binary, ternary type reaction objects should realistically belong in the `chemical_reactions` module.

**Reply by cpgr** [reply]
I think the regex parser in chemical reactions actions are similar to what is in crane. It too reads the reactions as a string, then adds all the appropriate objects.
